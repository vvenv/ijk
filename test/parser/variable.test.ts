import { describe, expect, test } from 'vitest';
import { parse } from './__helper';

test('basic', () => {
  expect(parse('{{ x }}')).toMatchSnapshot();
  expect(parse('{{= x }}')).toMatchSnapshot();
});

test('multiple', () => {
  expect(parse('{{ x }} and {{ y }}')).toMatchSnapshot();
});

test('duo', () => {
  expect(parse('{{ x }} and {{ x }}')).toMatchSnapshot();
});

test('nesting', () => {
  expect(parse('{{ x }}{{ x.y }}{{ x.y.z }}')).toMatchSnapshot();
});

describe('literal', () => {
  test('string', () => {
    expect(parse('{{ "*" }}')).toMatchSnapshot();
  });

  test('number', () => {
    expect(parse('{{ 1314 }}')).toMatchSnapshot();
    expect(parse('{{ 13.14 }}')).toMatchSnapshot();
  });

  test('boolean', () => {
    expect(parse('{{ true }}')).toMatchSnapshot();
    expect(parse('{{ false }}')).toMatchSnapshot();
  });

  test('array', () => {
    expect(parse('{{ [1, "1"] }}')).toMatchSnapshot();
  });

  test('object', () => {
    expect(parse('{{ { x: 1 } }}')).toMatchSnapshot();
  });

  test.skip('regexp', () => {
    expect(parse('{{ /\\d+/ }}')).toMatchSnapshot();
    expect(parse('{{ /{{.+?}}/gms }}')).toMatchSnapshot();
  });
});

describe('filter', () => {
  test('basic', () => {
    expect(parse('{{ x | upper }}')).toMatchSnapshot();
  });

  test('multiple', () => {
    expect(parse('{{ x | upper | lower }}')).toMatchSnapshot();
  });

  test('duo', () => {
    expect(parse('{{ x | upper }} and {{ x | upper }}')).toMatchSnapshot();
  });

  test('safe', () => {
    expect(parse('{{ x | safe }}')).toMatchSnapshot();
  });

  test('w/ params', () => {
    expect(parse(`{{ name | split "" }}`)).toMatchSnapshot();
  });
});

describe('expression', () => {
  test('arithmetic', () => {
    expect(
      parse('{{ x+2 }}{{ x-2 }}{{ x*2 }}{{ x/2 }}{{ x**2 }}'),
    ).toMatchSnapshot();
  });

  test('array/object member', () => {
    expect(parse('{{ x[2] }}{{ y["foo"] }}{{ z.bar }}')).toMatchSnapshot();
  });
});
