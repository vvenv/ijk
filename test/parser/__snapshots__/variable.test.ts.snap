// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(c.x);return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ x }}",
  },
}
`;

exports[`basic 2`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(c.= x);return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{= x }}",
  },
}
`;

exports[`duo 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(c.x);s+=" and ";s+=e(c.x);return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ x }} and {{ x }}",
  },
}
`;

exports[`expression > arithmetic 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(c.x+2);s+=e(c.x-2);s+=e(c.x*2);s+=e(c.x/2);s+=e(c.x**2);return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ x+2 }}{{ x-2 }}{{ x*2 }}{{ x/2 }}{{ x**2 }}",
  },
}
`;

exports[`expression > array/object member 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(c.x[2]);s+=e(c.y["foo"]);s+=e(c.z.bar);return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ x[2] }}{{ y["foo"] }}{{ z.bar }}",
  },
}
`;

exports[`filter > basic 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(f.upper(c.x));return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ x | upper }}",
  },
}
`;

exports[`filter > duo 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(f.upper(c.x));s+=" and ";s+=e(f.upper(c.x));return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ x | upper }} and {{ x | upper }}",
  },
}
`;

exports[`filter > multiple 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(f.lower(f.upper(c.x)));return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ x | upper | lower }}",
  },
}
`;

exports[`filter > safe 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(f.safe(c.x));return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ x | safe }}",
  },
}
`;

exports[`filter > w/ params 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(f.split(c.name,""));return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ name | split "" }}",
  },
}
`;

exports[`literal > array 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e([1, "1"]);return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ [1, "1"] }}",
  },
}
`;

exports[`literal > boolean 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(true);return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ true }}",
  },
}
`;

exports[`literal > boolean 2`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(false);return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ false }}",
  },
}
`;

exports[`literal > number 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(1314);return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ 1314 }}",
  },
}
`;

exports[`literal > number 2`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(13.14);return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ 13.14 }}",
  },
}
`;

exports[`literal > object 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e({ x: 1 });return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ { x: 1 } }}",
  },
}
`;

exports[`literal > string 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e("*");return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ "*" }}",
  },
}
`;

exports[`multiple 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(c.x);s+=" and ";s+=e(c.y);return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ x }} and {{ y }}",
  },
}
`;

exports[`nesting 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";s+=e(c.x);s+=e(c.x.y);s+=e(c.x.y.z);return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ x }}{{ x.y }}{{ x.y.z }}",
  },
}
`;
